src/Mafia/Package.hs

```diff
@@ -48,15 +57,38 @@ packageIdTuple (PackageId n v) =
 
 -- Extract name from `$name-$version`, but consider `unordered-containers-1.2.3`
 parsePackageId :: Text -> Maybe PackageId
-parsePackageId  =
+parsePackageId =
   let parser =
         PackageId
           <$> (PackageName . T.intercalate "-" . fmap T.pack <$> Parse.sepBy1 component (Parse.char '-'))
           <* Parse.char '-'
-          <*> parseVersion
+          <*> Version.parseVersion
           <* Parse.eof
-  in fmap fst . listToMaybe . reverse . Parse.readP_to_S parser . T.unpack
+  in parseLongestMatch parser
   where
     component = do
       cs <- Parse.munch1 Char.isAlphaNum
       if all Char.isDigit cs then Parse.pfail else return cs
+
+parseVersion :: Text -> Maybe Version
+parseVersion =
+  parseLongestMatch Version.parseVersion
+
+parseLongestMatch :: Parse.ReadP a -> Text -> Maybe a
+parseLongestMatch p =
```

Comment:
Yeah I'm tempted to rewrite the lot as an attoparsec parser just for consistency